cnmodel/data/_db.py

```diff
@@ -208,13 +213,27 @@ def add_table_data(name, row_key, col_key, data, **kwds):
     #print col_names
     #print row_names
     #print cells
+
+    changes = [] # a list of parameters that are changed if we are rewriting a table
     for i,row in enumerate(row_names):
         for j,col in enumerate(col_names):
             kwds[row_key] = row
             kwds[col_key] = col
-            setval(cells[i][j], name, **kwds)
+            oldval = setval(cells[i][j], name, **kwds)
+            if oldval is not None and oldval != cells[i][j]:
+                changes.append({'name': name, 'row': row, 'col': col, 'new': cells[i][j], 'old': oldval})
```

Comment:
There's a possibility for some of the items in `changes` to be ambiguous since it does not include all of `**kwds` (setting two different keys could result in two identical items in `changes`). Maybe:
```
    key = mk_key(name, **kwds)
    changes.append({'key': key, new: cells[i][j], old: oldval})
```